### DISCUSSION TITLE
Best way to work with $record

### DISCUSSION THREAD

**Discussion by labiangashi-prosperoware** [codaxy/cxjs#904] [Q&A] [answered]
Hi everyone,

I am having some issues working with the `$record` attribute inside Repeaters...

I have a form (created with a `Repeater` widget) and it has many records (represented by `$record`) but I am having some issues controlling them correctly.

For example, when I press a button next to the record, I want that specific record to be chosen and represented in `$record`. Am I doing it incorrectly or am I missing something crucial? Sometimes, that `$record` is not chosen and something else is instead. What is the correct and effective way to select the records I want inside the repeater widget?

Thanks.

**Comment by mstijak** [comment]
Yes, this is supported. You can use the second argument in the `onClick` callback method to retrieve the store for that particular record.

```jsx
<Button onClick={(event, instance) => { console.log(instance.store.get('$record')); }} />
```
It's common to use destructuring to make it a bit shorter.

```jsx
<Button onClick={(event, { store }) => { console.log(store.get('$record')); }} />
```

If you don't use the store provided with the callback, `$record` will always point to the last record rendered.

**Reply by labiangashi-prosperoware** [reply]
I found the issue and don't know whether this is a bug (might be).

So here's the thing, I have (in this case) three form fields with each one having a tooltip attached to them. All the tooltips have a `alwaysVisible` property set to true and the form fields are lookups. 
Here's the initial DOM:
![image](https://user-images.githubusercontent.com/73879945/141769577-83e272d7-7a3e-4d8c-92ab-ccb011e4ea44.png)

After I open the lookup, one of those absolute positioned elements disappeared and the tooltip is gone forever....
![image](https://user-images.githubusercontent.com/73879945/141769704-f15e2273-9031-4913-aa46-737fc4a2bfd1.png)

Is there a workaround for this and should this be reported as an issue? (Eventhough it looks like an edge case to me but it can happen)

**Reply by mstijak** [reply]
Please provide a [fiddle](https://fiddle.cxjs.io) reproducing the issue. It's very hard to figure it out based on the description you provided.

**Comment by labiangashi-prosperoware** [comment]
@mstijak 
https://fiddle.cxjs.io/?f=q6oagJ0T

It apparently is happening when the field is required, if it is and it triggers the validation, the tooltip never shows again.

**Reply by labiangashi-prosperoware** [reply]
That kind of makes sense since the validation tooltip overrides the one on the component so it is a little difficult to maintain both... Do you have any suggestions for a workaround though?

**Comment by mstijak** [comment, marked as answer]
Ah, I know what happens. One widget can have only one tooltip, so the error
tooltip takes over when the field becomes invalid.

Maybe you can wrap the field in a div and put the "alwaysVisible" tooltip
on that div. You can use a LabeledContainer -
https://docs.cxjs.io/widgets/labeled-containers if you encounter problems
with labels.

**Reply by labiangashi-prosperoware** [reply]
Yep, that solved the issue! I initially thought it had to do with the lookup field itself but then found out it's all about the tooltip and the `alwaysVisible` property.

Wrapping them both around a `<div>` fixed the issue, I'll mark this as answered, thanks!
